runtime/gc_base/ClassLoaderManager.cpp

```diff
@@ -489,23 +496,48 @@ MM_ClassLoaderManager::cleanUpSegmentsInAnonymousClassLoader(MM_EnvironmentBase
 				/* Anonymous classes expected to be allocated one per segment */
 				Assert_MM_true(NULL == classHeapIterator.nextClass());
 
+				totalRAMClasses += 1;
+
 				if (J9AccClassDying == (J9CLASS_FLAGS(clazz) & J9AccClassDying)) {
-					/* TODO replace this to better algorithm */
-					/* Try to find ROM class for unloading anonymous RAM class if it is not an array */
-					if (!_extensions->objectModel.isIndexable(clazz)) {
+
+					deadRAMClasses += 1;
+
+					/* Try to find ROM class for unloading anonymous RAM class if it is possible to unload */
+					if (isROMClassUnloadable(clazz)) {
 						J9ROMClass *romClass = clazz->romClass;
-						if (NULL != romClass) {
-							J9MemorySegment **previousSegmentPointerROM = &_javaVM->anonClassLoader->classSegments;
-							J9MemorySegment *segmentROM = *previousSegmentPointerROM;
 
+						unloadableROMClasses += 1;
+
+						/* Check next segment, there is a high probability it is what we are looking for */
+						if ((NULL != nextSegment)
+								&& (MEMORY_TYPE_ROM_CLASS == (nextSegment->type & MEMORY_TYPE_ROM_CLASS))
+								&& ((J9ROMClass *)nextSegment->heapBase == romClass)) {
+
+							foundROMClassesFast += 1;
+
+							/* Yes, next segment is the ROM segment to be freed */
+							J9MemorySegment *segmentToFree = nextSegment;
+
+							/* correct pre-cached next for RAM iteration cycle */
+							nextSegment = segmentToFree->nextSegmentInClassLoader;
```

Comment:
add a comment:
- either here why we are not removing from the list (while slow path does it)
- or better, at the place where we remove it (in the existing code), that we might be removing both RAM and ROM in one shot

Reply:
added